File: ros/move_joints.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#
"""
joint_control_sim_test.py

Пример управления манипулятором Rooky в ROS (режим симуляции).

"""
# Импортируем необходимые библиотеки
# библиотека работы с ROS
import rospy

# Данный тип сообщений необходим для trajectory_msgs
from std_msgs.msg import *

# Сообщения для описания траектории движения
from trajectory_msgs.msg import *

class ControlJoints:
    def __init__(self, arm_type):
        # Сохраним тип Rooky
        self._arm_type = arm_type
        self._positions = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

        # Получим частоту от сервера параметров ROS. 
        # Если не удалось, по умолчанию 20Гц
        self._rate = rospy.get_param('~rate', 20)

        # Список для движения всех суставов
        self._joint_names = [str(arm_type) + '_arm_1_joint', 
                             str(arm_type) + '_arm_2_joint',
                             str(arm_type) + '_arm_3_joint', 
                             str(arm_type) + '_arm_4_joint',
                             str(arm_type) + '_arm_5_joint', 
                             str(arm_type) + '_arm_6_joint',
                             str(arm_type) + '_arm_7_joint']

        # Подключимся к топику как Publisher
        # Тип сообщений - JointTrajectory
        # Максимальное число сообщений в очереди на отправку - 1
        self._cmd_pub = rospy.Publisher(str(arm_type) + '_arm_controller/command',
                                        JointTrajectory,
                                        queue_size=1)

        # Задержка на 1 секунду
        rospy.sleep(1)

        # Время движения до указанной точки в секундах
        interval_to_point = 1
        
        self._positions = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
        self.move_all_joints(interval_to_point)

        rospy.sleep(2.0)

    def spinOnce(self):
         # Время движения до указанной точки в секундах
        interval_to_point = 2
        self._positions = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
        self.move_all_joints(interval_to_point)
        rospy.sleep(5)

        self._positions = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
        self.move_all_joints(interval_to_point)
        rospy.sleep(5)

        # Движения одного сустава
        self.move_joint(str(self._arm_type) + '_arm_1_joint', 1.0, 3)
        rospy.sleep(5)

    def rate(self):
        return self._rate

    # Функция отправки в желаемую позицию всех суставов Rooky
    def move_all_joints(self, interval_to_point=0.1):
        # Создадим объект траектории
        traj = JointTrajectory()

        # Назначим имена суставов
        traj.joint_names = self._joint_names

        # Создадим объект точки, которой должен достичь манипулятор Rooky.
        point = JointTrajectoryPoint()

        # Назначим позиции для движения каждого сустава
        point.positions = self._positions

        # Укажем как долго должны двигаться манипулятор до указанного положения
        point.time_from_start = rospy.Duration(interval_to_point)
        traj.points.append(point)

        # Отправим сообщение в топик
        self._cmd_pub.publish(traj)

    # Функция отправки в желаемую позицию одного сустава Rooky
    def move_joint(self, joint_name, position, interval_to_point=0.1):
        traj = JointTrajectory()
        traj.joint_names.append(joint_name)
        point = JointTrajectoryPoint()

        point.positions.append(position)

        point.time_from_start = rospy.Duration(interval_to_point)
        traj.points.append(point)

        self._cmd_pub.publish(traj)


rospy.init_node('joint_control_sim_test')

# Создадим узел ROS
node = ControlJoints("left")
while not rospy.is_shutdown():
    node.spinOnce()
    rospy.Rate(node.rate()).sleep()

File: ros/cpp_examples_ws/src/examples/src/move_joints.cpp
#include "ros/ros.h"
#include "trajectory_msgs/JointTrajectory.h"
#include <string>
#include <vector>

#define JOINTS_COUNT 7

enum RookySide
{
	LEFT = 0,
	RIGHT
};

int main(int argc, char **argv)
{
	// Определим тип Rooky
	RookySide rooky_side = LEFT;

	// Инициализируем ROS с указанием названия узла
	ros::init(argc, argv, "move_joints");
	ros::NodeHandle nh;
	ros::Rate loop_rate(20);

	// В зависимости от типа Rooky определим наименование топика
	// в который необходимо отправлять сообщение
	std::string controller_name;
	if (rooky_side == RIGHT)
		controller_name = "right_arm_controller/command";
	else
		controller_name = "left_arm_controller/command";

	// Определим publisher с очередью в 1 сообщение
	ros::Publisher msg = nh.advertise<trajectory_msgs::JointTrajectory>(controller_name.c_str(), 1);

	// Создадим сообщение которое будем отправлять в топик
	// Данное сообщение будет содержать положение сразу всех суставов
	trajectory_msgs::JointTrajectory joints_state;
	joints_state.joint_names.resize(JOINTS_COUNT);
	for (int i = 1; i < JOINTS_COUNT + 1; i++)
	{
		// Построим имя сустава, 
		// например строка с именем должна выглядеть так:
		// "left_arm_1_joint"
		std::string joint_name;
		if (rooky_side == RIGHT)
			joint_name += "right";
		else
			joint_name += "left";

		joint_name += "_arm_";

		// Переведем int в ASCII цифру
		joint_name += boost::lexical_cast<std::string>(i); 
		joint_name += "_joint";

		// Добавим имя сустава в сообщение
		joints_state.joint_names[i - 1] = joint_name;
	}

	// Создадим описание положений - Точки
	std::vector<trajectory_msgs::JointTrajectoryPoint> points(1);
	while (ros::ok())
	{
		// Не будем задавать скорость движения
		points[0].velocities.resize(0);

		// Задаим время, которое должна двигаться Rooky до указанного положения
		points[0].time_from_start.sec = 1;

		points[0].positions.resize(JOINTS_COUNT);

		// Последовательно запишем позиции для каждого узла
		for (int i = 0; i < JOINTS_COUNT; i++)
		{
			// Вначале все суставы двигаем на 1 радиан
			// Потом все суставы возвращаем в 0
			if (points[0].positions[i] != 0)
				points[0].positions[i] = 0;
			else
				points[0].positions[i] = 1;
		}

		joints_state.points = points;

		// Опубликуем сообщение в топик
		msg.publish(joints_state);

		// Сделаем паузу на 5 секунд
		ros::Duration(5).sleep();
		ros::spinOnce();
		loop_rate.sleep();
	}
	return 0;
}
